lib/SGN/Controller/solGS/pca.pm

```diff
@@ -160,64 +174,124 @@ sub pca_result :Path('/pca/result/') Args() {
- 
-    $c->stash->{file_id} = $file_id;
     
-    $self->pca_scores_file($c);
-    my $pca_scores_file = $c->stash->{pca_scores_file};
+    $c->stash->{file_id} = $file_id;
 
-    $self->pca_variance_file($c);
-    my $pca_variance_file = $c->stash->{pca_variance_file};
- 
-    my $ret->{status} = 'PCA analysis failed.';
-    my $pca_scores;
-    my $pca_variances;
-    
-    if( !-s $pca_scores_file)
-    {	
+    my $ret->{status} = 'PCA analysis failed';
+   
+    if (!$self->check_pca_output($c)) 
+    {
+	$self->create_pca_genotype_data($c);
+	
 	if (!$c->stash->{genotype_files_list} && !$c->stash->{genotype_file}) 
 	{	  
 	    $ret->{status} = 'There is no genotype data. Aborted PCA analysis.';                
 	}
 	else 
-	{
-	    $self->run_pca($c);	    
-	}	
+	{ 
+	    $self->run_pca($c);
+	    $self->format_pca_output($c);
+	}
+	
     }
-    
-    $pca_scores    = $c->controller('solGS::solGS')->convert_to_arrayref_of_arrays($c, $pca_scores_file);
-    $pca_variances = $c->controller('solGS::solGS')->convert_to_arrayref_of_arrays($c, $pca_variance_file);
 
-    my $host = $c->req->base;
+    $ret = $c->stash->{formatted_pca_output};
+    $ret->{pop_id} = $c->stash->{pop_id};
+    $ret->{trials_names} = $trials_names;
+   
+    $ret = to_json($ret);
+       
+    $c->res->content_type('application/json');
+    $c->res->body($ret);    
 
-    if ( $host !~ /localhost/)
+}
+
+sub check_pca_output {
+    my ($self, $c) = @_;
+
+    my $file_id = $c->stash->{file_id};
+
+    if ($file_id)
     {
-	$host =~ s/:\d+//; 
-	$host =~ s/http\w?/https/;
+	$self->format_pca_output($c);
+	my $pca_output = $c->stash->{formatted_pca_output};
+
+	if (ref($pca_output) eq 'HASH') 
+	{
+	    return 1;
+	}
+	else
+	{
+	    return 0;
+	}
     }
-    
-    my $output_link = $host . 'pca/analysis/' . $pop_id;
 
-    if ($pca_scores)
+}
+
+
+sub format_pca_output {
+    my ($self, $c) = @_;
+
+    my $file_id = $c->stash->{file_id};
+
+    if ($file_id)
     {
-        $ret->{pca_scores} = $pca_scores;
-	$ret->{pca_variances} = $pca_variances;
-        $ret->{status} = 'success';  
-	$ret->{pop_id} = $c->stash->{pop_id};# if $list_type eq 'trials';
-	$ret->{trials_names} = $c->stash->{trials_names};
-	$ret->{output_link}  = $output_link;
-    }
+	$self->pca_scores_file($c);
+	my $pca_scores_file = $c->stash->{pca_scores_file};
 
-    $ret = to_json($ret);
-       
-    $c->res->content_type('application/json');
-    $c->res->body($ret);    
+	$self->pca_variance_file($c);
+	my $pca_variance_file = $c->stash->{pca_variance_file};
+	
+	if ( -s $pca_scores_file && -s $pca_variance_file)
+	{
+	    my $ret->{status} = undef;
+	    my $pca_scores    = $c->controller('solGS::solGS')->convert_to_arrayref_of_arrays($c, $pca_scores_file);
+	    my $pca_variances = $c->controller('solGS::solGS')->convert_to_arrayref_of_arrays($c, $pca_variance_file);
 
-}
+	    my $host = $c->req->base;
 
+	    if ( $host !~ /localhost/)
+	    {
+		$host =~ s/:\d+//; 
```

Comment:
There should never be a check for localhost in the code, and I am not sure why this is needed in this case?

Reply:
that is just for when working on my localhost.

Reply:
my localhost has the port number 3000 (http://localhost:3000).  On the test site and production site the 8080 port number is include when you do host and referer request, which creates links that break.